src/stubgen.py

```diff
@@ -81,11 +82,63 @@ class and repeatedly call ``.put()`` to register modules or contents within the
 ]
 # fmt: on
 
+# (name, desired_as_name) -> actual_as_name
+ImportDict = Dict[Tuple[Optional[str], Optional[str]], Optional[str]]
+# package_name -> ((name, desired_as_name) -> actual_as_name)
+PackagesDict = Dict[str, ImportDict]
+
+# (signature_str, doc_str, default_arg_1, default_arg_2, ...)
+# We could type as Tuple[Optional[str], Optional[str], *Tuple[Any, ...]]
+# but this is only valid in python >= 3.11
+NbSignature = Tuple[Optional[str], Optional[str], Any]
+NbGetterSetterSignature = Tuple[str, str]
+
+
+class NbObject(Protocol):
+    __module__: Literal["nanobind"]
+    __name__: str
+    __nb_signature__: Tuple[NbSignature, ...]
+    __bases__: Tuple[Any, ...]
+
+
+class NbFunction(Protocol):
+    __module__: Literal["nanobind"]
+    __name__: Literal["nb_func", "nb_method"]
+    __nb_signature__: Tuple[NbSignature, ...]
+    __func__: "NbFunction"
+    __call__: Callable[..., Any]
+
+
+@typing.runtime_checkable
+class NbGetter(Protocol):
+    __nb_signature__: Tuple[NbGetterSetterSignature, ...]
+    __call__: Callable[[Any], Any]
```

Comment:
Is it important to have all of these things listed here? For example, I'm pretty sure we are not calling the getter.

I think there is a more fundamental question: are we trying to create some kind of object abstraction layer for nanobind in general? Then stubgen probably isn't the right place for it. I would keep it minimally short and simple, which has also been the goal for other parts of this file.

Reply:
See also this.